Add product form extensibility tutorials

louwie17 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Add two tutorials, one on using the generic fields and the second one on writing a custom field.
**Note:** Some of the links that link to the new markdown files are pointing at `trunk` currently, so they may not work until merged.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Read through the documentation to make sure everything looks ok
2. Walk through the tutorials to see if you can follow it and it works.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

louwie17 committed:
Start on the generic fields tutorial

louwie17 committed:
Add custom field tutorial

louwie17 committed:
Some small updates

louwie17 committed:
Address linting

louwie17 committed:
Minor updates

louwie17 committed:
Update manifest

nathanss commented on docs/product-editor-development/how-to-guides/custom-field-tutorial.md:
```diff
@@ -0,0 +1,263 @@
+To get started we would recommend reading through the [fundamentals of block development docs](https://developer.wordpress.org/block-editor/getting-started/fundamentals/) in WordPress. This gives a good overview of working with blocks, the block structure, and the [JavaScript build process](https://developer.wordpress.org/block-editor/getting-started/fundamentals/javascript-in-the-block-editor/).
+
+This tutorial will use vanilla JavaScript to render a new field in the product form for those that already have a plugin and may not have a JavaScript build process set up yet.
+If you wanted to create a plugin from scratch with the necessary build tools I recommend using the `@wordpress/create-block` script, we also have a specific template for the product form: [README](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/create-product-editor-block/README.md).
```
```suggestion
If you want to create a plugin from scratch with the necessary build tools, we recommend using the `@wordpress/create-block` script. We also have a specific template for the product form: [README](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/create-product-editor-block/README.md).
```

nathanss commented on docs/product-editor-development/how-to-guides/custom-field-tutorial.md:
```diff
@@ -0,0 +1,263 @@
+
+Adding and registering our custom field is very similar as creating a brand new block.
+
+Inside a new folder within your plugin lets create a `block.json` file with the contents below. The only main difference between this `block.json` and a `block.json` for the site editor is that we will set `supports.inserter` to false, so it doesn't show up there. We will also be registering this slightly different.
```
```suggestion
Inside a new folder within your plugin, let's create a `block.json` file with the contents below. The only main difference between this `block.json` and a `block.json` for the site editor is that we will set `supports.inserter` to false, so it doesn't show up there. We will also be registering this slightly different.
```

nathanss commented on docs/product-editor-development/how-to-guides/custom-field-tutorial.md:
````diff
@@ -0,0 +1,263 @@
+}
+```
+
+In the same directory create a `index.js` file, which we can keep simple by just outputting a hello world.
````
```suggestion
In the same directory, create a `index.js` file, which we can keep simple by just outputting a hello world.
```

nathanss commented on docs/product-editor-development/how-to-guides/custom-field-tutorial.md:
````diff
@@ -0,0 +1,263 @@
+} )( window.wp );
+```
+
+In react:
````
```suggestion
In React:
```

nathanss commented on docs/product-editor-development/how-to-guides/custom-field-tutorial.md:
```diff
@@ -0,0 +1,263 @@
+import { registerBlockType } from '@wordpress/blocks';
+
+function Edit() {
+    return <p>Hello World (from the editor).<p>;
```
```suggestion
    return <p>Hello World (from the editor).</p>;
```

nathanss commented on docs/product-editor-development/how-to-guides/custom-field-tutorial.md:
````diff
@@ -0,0 +1,263 @@
+} );
+```
+
+Lastly in order to make this work the block registration needs to know about the JavaScript dependencies, we can do so by adding a `index.asset.php` file with the below contents:
````
```suggestion
Lastly, in order to make this work the block registration needs to know about the JavaScript dependencies, we can do so by adding a `index.asset.php` file with the below contents:
```

nathanss commented on docs/product-editor-development/how-to-guides/custom-field-tutorial.md:
```diff
@@ -0,0 +1,263 @@
+
+We can add it to the product form by hooking into the `woocommerce_layout_template_after_instantiation` action ( see [block addition and removal](https://github.com/woocommerce/woocommerce/blob/trunk/docs/product-editor-development/block-template-lifecycle.md#block-addition-and-removal) ).
+
+What I did was the following ( see [here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/README.md#usage) for more related functions ):
```
Can we avoid the use of "I" here somehow?

nathanss commented on docs/product-editor-development/how-to-guides/custom-field-tutorial.md:
```diff
@@ -0,0 +1,263 @@
+
+We recommend using components from `@wordpress/components` as this will also keep the styling consistent. We will use the [ComboboxControl](https://wordpress.github.io/gutenberg/?path=/docs/components-comboboxcontrol--docs) core component in this field.
+
+We can add it to our `edit` function pretty easily by making use of `wp.components`. I will also add a constant for the filter options.
```
Same as before, I would try to avoid "I will also add", "I also added", etc.

nathanss commented on docs/product-editor-development/how-to-guides/generic-fields-tutorial.md:
```diff
@@ -0,0 +1,58 @@
+
+Using a generic block is pretty easy. We have created an template API that allows you to add new fields, the API refers to them as `blocks`. There are a couple actions that allow us to interact with these templates. There is the `woocommerce_layout_template_after_instantiation` that is triggered when a new template is registered. There are also other actions triggered when a specific field/block  ...
+
+Lets say we want to add something to the basic details section, we can do so by making use of the above mentioned hook:
```
```suggestion
Let's say we want to add something to the basic details section, we can do so by making use of the above mentioned hook:
```

nathanss reviewed (changes requested):
Added some suggestions. Overall, it looks good! Nice job!

louwie17 committed:
Address PR feedback, mostly formatting changes

louwie17 commented:
> Added some suggestions. Overall, it looks good! Nice job!

Thanks for the review @nathanss, I have addressed all your suggestions as part of this commit: https://github.com/woocommerce/woocommerce/pull/50137/commits/a58739f125580f23432cc1447b3a1b8492acf615
